contracts/cw20-ics20/src/contract.rs

```diff
@@ -186,14 +192,54 @@ pub fn execute_allow(
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
-pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
-    let version = get_contract_version(deps.storage)?;
-    if version.contract != CONTRACT_NAME {
+pub fn migrate(mut deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
+    let version: Version = CONTRACT_VERSION.parse().map_err(from_semver)?;
+    let stored = get_contract_version(deps.storage)?;
+    let storage_version: Version = stored.version.parse().map_err(from_semver)?;
+
+    // First, ensure we are working from an equal or older version of this contract
+    // wrong type
+    if CONTRACT_NAME != stored.contract {
         return Err(ContractError::CannotMigrate {
-            previous_contract: version.contract,
+            previous_contract: stored.contract,
         });
     }
-    Ok(Response::default())
+    // existing one is newer
+    if storage_version > version {
+        return Err(ContractError::CannotMigrateVersion {
+            previous_version: stored.version,
+        });
+    }
+
+    // Then, run the proper migration
+    // unsupported old version
+    if storage_version < "0.11.1".parse().map_err(from_semver)? {
+        return Err(ContractError::CannotMigrateVersion {
+            previous_version: stored.version,
+        });
+    }
+    // for 0.12.0-alpha1 or earlier, migrate from the config.gov_contract to ADMIN
+    if storage_version <= "0.12.0-alpha1".parse().map_err(from_semver)? {
+        // Do v1 migration
+        let old_config = crate::migrations::v1::CONFIG.load(deps.storage)?;
+        ADMIN.set(deps.branch(), Some(old_config.gov_contract))?;
+        let config = Config {
+            default_timeout: old_config.default_timeout,
+        };
+        CONFIG.save(deps.storage, &config)?;
+    }
+    // otherwise no migration (yet)
+
+    if storage_version < version {
+        // we don't need to save anything if migrating from the same version
+        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    }
+
+    Ok(Response::new())
+}
```

Comment:
Not for now, but I wonder how much of this logic can be turned into a helper function.

Reply:
some of this is in poe-contracts/tg-utils

but matching on versions would be a good thing to abstract out (later, when we have some good examples/experience)